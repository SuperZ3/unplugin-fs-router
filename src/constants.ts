import { SegementType } from "./types";

export const MODULE_PATH = "unplugin-fs-router/routes";

export const VIRTUAL_ID_PREFIX = "\0";

export const VIRTUAL_ID = VIRTUAL_ID_PREFIX + MODULE_PATH

export enum PkgType {
    react = 'react',
    vue = 'vue'
}

export const PkgTypes = [PkgType.react, PkgType.vue];

export const SegmentTypeRexp = {
    [SegementType.Base]: /^([a-zA-Z0-9_-]+[a-zA-Z0-9_\-\.]*[a-zA-Z0-9_-]+)$/,
    [SegementType.Dynamic]: /^\[([a-zA-Z0-9_-]+)\]$/,
    [SegementType.Children]: /^([a-zA-Z0-9_-]+)\/$/,
    [SegementType.Group]: /^(\([a-zA-Z0-9_-]+\))$/,
    [SegementType.CatchAll]: /^\[\.{3}([a-zA-Z0-9_-]+)\]$/,
};

export const DefaultExportRexp = /export\sdefault\s(class\s|function\s|const\s)?([a-zA-Z0-9_-]+)\s/

export const DefaultReactExtensions = [".js", ".jsx", ".ts", ".tsx"]

export const DefaultVueExtensions = [".js", ".ts", ".vue"]

export const RootTreeNodeKey = "$_ROOT_TREE_NODE_KEY_$"