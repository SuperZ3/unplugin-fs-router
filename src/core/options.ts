import path from "path";
import {
	DefaultReactExtensions,
	DefaultVueExtensions,
	PkgType,
	PkgTypes,
} from "../constants";
import { Options, UserOptions } from "../types";
import { isArrray } from "../utils";
import { existsSync, readFileSync } from "fs";

export const DefaultOptions: Options = {
	root: process.cwd(),
	rootPaths: ["src/pages"],
	extensions: DefaultReactExtensions,
	exclude: [],
	pkgType: PkgType.react,
};

// Scan package.json to verify project is React„ÄÅVue...
function resolvePkgType(
	userPkgType: PkgType.react | PkgType.vue | undefined,
	root: string
): PkgType {
	if (userPkgType) return userPkgType;

	const pkgJSONPath = path.resolve(root, "package.json");
	if (existsSync(pkgJSONPath)) {
		const pkgJSON = readFileSync(pkgJSONPath, { encoding: "utf-8" });
		const pkgContent = JSON.parse(pkgJSON);
		const dependencies = pkgContent?.dependencies ?? {};
		const devDependencies = pkgContent?.devDependencies ?? {};
		const peerDependencies = pkgContent?.peerDependencies ?? {};
		for (let t of PkgTypes.values()) {
			if (t in dependencies || t in devDependencies || t in peerDependencies) {
				return t;
			}
		}
	}
	return PkgType.react;
}

function mergerootPaths(path: undefined | string | string[]): string[] {
	if (path === undefined) {
		return DefaultOptions.rootPaths;
	}
	return isArrray(path) ? path : [path];
}

function resolveExtensions(
	pkgType: PkgType.react | PkgType.vue,
	extensions: string[] | undefined
): string[] {
	if (extensions !== undefined) {
		return extensions;
	}
	if (pkgType === PkgType.react) {
		return DefaultReactExtensions;
	} else if (pkgType === PkgType.vue) {
		return DefaultVueExtensions;
	}
	return DefaultOptions.extensions
}

export function getContextOptions(
	userOptions: UserOptions | undefined = {}
): Options {
	const root = userOptions?.root ?? DefaultOptions.root;
	const rootPaths = mergerootPaths(userOptions?.rootPaths);
	const pkgType = resolvePkgType(userOptions?.pkgType, root);
	const extensions = resolveExtensions(pkgType, userOptions?.extensions);
	return {
		...DefaultOptions,
		...userOptions,
		extensions,
		rootPaths,
		pkgType,
	};
}
