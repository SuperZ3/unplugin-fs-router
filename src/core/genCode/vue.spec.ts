import path from "path";
import { describe, expect, it } from "vitest";
import Context from "../context";
import { DefaultOptions } from "../options";
import { DefaultVueExtensions } from "../../constants";


describe("react routes", () => {
	it("match snapshot", async () => {
		const options = {
			...DefaultOptions,
			rootPaths: ["/examples/vue/src/pages"],
			extensions: DefaultVueExtensions,
		};
		const ctx = new Context(options);
		await ctx.start();
		expect(ctx.genRoutes()).toMatchSnapshot();
	});
});
