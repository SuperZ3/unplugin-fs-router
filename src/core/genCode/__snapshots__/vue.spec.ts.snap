// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`react routes > match snapshot 1`] = `
"import React, { createElement } from 'react';


const Routes = [

]
export default Routes;"
`;
